Modules/Remotes/CloneRepositoryDlg.cpp

```diff
@@ -0,0 +1,258 @@
+        {
+            okay = false;
+        }
+    }
+
+    setAcceptable( okay );
+}
+
+void CloneDlg::accept()
+{
+    Git::CloneOperation* clone = new Git::CloneOperation( this );
+    QString repoName  = QUrl(mCloneWdgt->txtUrl->text() )
+                        .adjusted(QUrl::NormalizePathSegments |
+                                  QUrl::StripTrailingSlash ).toString();
+    QString targetDir = QUrl(mCloneWdgt->txtPath->text() )
+                        .adjusted(QUrl::NormalizePathSegments |
+                                  QUrl::StripTrailingSlash ).toString();
+
+    if ( mCloneWdgt->chkAppendRepoName->isChecked() )
+    {
+        targetDir += QStringLiteral("/%1")
+                    .arg( QUrl( repoName ).fileName() );
+    }
+
+    clone->setBackgroundMode( true );
+    clone->setUrl( repoName );
+    clone->setPath( targetDir );
+    clone->setRemoteAlias( mCloneWdgt->txtRemoteAlias->text() );
+
+    if ( mCloneOptsWdgt )
+    {
+        clone->setBare( mCloneOptsWdgt->mCloneMode == CloneOptionsWdgt::cmBare );
+        clone->setReference( mCloneOptsWdgt->txtBranch->text() );
+        clone->setDepth( mCloneOptsWdgt->txtCloneDepth->value() );
+    }
+
+    mProgress = new ProgressDlg;
+
+    // TODO: implement a ProgressDlg::minimumDuration
+    mProgress->show();
+
+    if( repoName.endsWith( QStringLiteral( ".git" ) ) )
+        repoName = repoName.left( repoName.length() - 4 );
+
+    if( repoName.lastIndexOf( QChar( L'/' ) ) != -1 )
+        repoName.remove( 0, repoName.lastIndexOf( QChar( L'/' ) ) + 1 );
+
+    ProgressDlg::StepInfo::List steps;
+    steps << ProgressDlg::StepInfo{ QStringLiteral("transfer"), tr("Download Git objects.") }
+          << ProgressDlg::StepInfo{ QStringLiteral("index"), tr("Add objects to Git index.") };
+
+    if (!clone->bare()) {
+          steps << ProgressDlg::StepInfo{ QStringLiteral("checkout"),
+                   tr("Checkout the worktree.") };
+          connect( clone, &Git::CloneOperation::checkoutProgress,
+                   this,  &CloneDlg::onCheckoutProgress );
+    }
+
+    mProgress->addActivity(tr("Clone <b>%1</b> to <b>%2</b>")
+                           .arg(repoName).arg(targetDir), clone, steps);
+
+    connect( clone, &Git::CloneOperation::finished,
+             this,  &CloneDlg::rootCloneFinished );
+    connect( clone, &Git::CloneOperation::transportProgress,
+             this,  &CloneDlg::onTransportProgress );
+
+    clone->execute();
+}
+
+void CloneDlg::rootCloneFinished()
+{
+    Git::BaseOperation* operation = static_cast<Git::BaseOperation*>( sender() );
+    Q_ASSERT( operation );
+
+    if ( operation->result() ) {
+        mProgress->finished(operation);
+    }
+    else {
+        mProgress->setError(operation, operation->result().errorText());
+    }
+
+}
+
+void CloneDlg::onCheckoutProgress(const QString& fileName, quint64 totalFiles,
+                                  quint64 completedFiles)
+{
+    QObject* s = sender();
```

Comment:
You simply cannot assume that a delayed slot's sender() is actually still valid. But it's not worth it, since we'll be doing this a  bit differently soon.